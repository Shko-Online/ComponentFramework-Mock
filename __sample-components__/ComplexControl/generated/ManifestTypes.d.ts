/*
    Copyright (c) 2022 Betim Beja and Shko Online LLC
    Licensed under the MIT license.
*/

// Define IInputs and IOutputs Type. They should match with ControlManifest.
export interface IInputs {
    inputProperty: ComponentFramework.PropertyTypes.TwoOptionsProperty;
    boundProperty: ComponentFramework.PropertyTypes.TwoOptionsProperty;
}
export interface IOutputs {
    boundProperty?: boolean;
    outputProperty?: any;
}
