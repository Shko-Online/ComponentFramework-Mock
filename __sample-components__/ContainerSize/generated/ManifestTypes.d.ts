/*
    Copyright (c) 2022 Betim Beja and Shko Online LLC
    Licensed under the MIT license.
*/

// Define IInputs and IOutputs Type. They should match with ControlManifest.
export interface IInputs {
    trackContainer: ComponentFramework.PropertyTypes.TwoOptionsProperty;
}
export interface IOutputs {
    trackContainer?: boolean;
}
