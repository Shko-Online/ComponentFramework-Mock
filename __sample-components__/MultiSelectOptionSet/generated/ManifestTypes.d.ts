/*
    Copyright (c) 2022 Betim Beja and Shko Online LLC
    Licensed under the MIT license.
*/

// Define IInputs and IOutputs Type. They should match with ControlManifest.
export interface IInputs {
    selection: ComponentFramework.PropertyTypes.MultiSelectOptionSetProperty;
}
export interface IOutputs {
    selection?: number[];
}
