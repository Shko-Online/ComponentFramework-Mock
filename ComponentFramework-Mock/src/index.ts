/*
Copyright (c) 2022 Betim Beja and Shko Online LLC
Licensed under the MIT license.
*/

/// <reference types="powerapps-component-framework" />
/// <reference path="./global.d.ts" />

export type { MockToRaw, PropertyMap, PropertyToMock } from './ComponentFramework-Mock';
export type { MockGenerator } from './ComponentFramework-Mock-Generator';
export type { ShkoOnline } from './ShkoOnline';

export {
    AttributeMetadataCollection,
    AttributeType,
    ClientMock,
    ContextMock,
    DataSetMock,
    DateTimeMetadataMock,
    DateTimePropertyMock,
    DecimalNumberMetadataMock,
    DecimalNumberPropertyMock,
    DeviceMock,
    EntityRecordMock,
    EnumPropertyMock,
    FactoryMock,
    FilteringMock,
    FormattingMock,
    LinkingMock,
    LookupMetadataMock,
    LookupPropertyMock,
    MetadataMock,
    ModeMock,
    MultiSelectOptionSetPropertyMock,
    NavigationMock,
    NumberMetadataMock,
    NumberPropertyMock,
    OptionMetadataMock,
    OptionSetMetadataMock,
    OptionSetPropertyMock,
    PagingMock,
    PopupServiceMock,
    PropertyMock,
    ResourcesMock,
    StringMetadataMock,
    StringPropertyMock,
    TwoOptionMetadataMock,
    TwoOptionsPropertyMock,
    UserSettingsMock,
    UtilityMock,
    WebApiMock,
    WholeNumberMetadataMock,
    WholeNumberPropertyMock,
} from './ComponentFramework-Mock';

export {
    ComponentFrameworkMockGeneratorReact,
    ComponentFrameworkMockGenerator,
    MetadataDB,
    ReactResizeObserver,
    mockGetEntityMetadata,
    mockRefreshParameters,
    mockSetControlResource,
    mockSetControlState,
} from './ComponentFramework-Mock-Generator';

export { AttributeMetadataGenerator, arrayEqual } from './utils';
