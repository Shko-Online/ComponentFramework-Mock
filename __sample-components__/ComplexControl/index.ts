import type { IInputs, IOutputs } from './generated/ManifestTypes';

export class ComplexControl implements ComponentFramework.StandardControl<IInputs, IOutputs> {
    private boundProperty: boolean;
    private inputProperty: boolean;
    private notifyOutputChanged: () => void;
    private updateCount = 0;
    private container: HTMLDivElement;
    /**
     * Empty constructor.
     */
    constructor() {}

    public async getOutputSchema(context: ComponentFramework.Context<IInputs>): Promise<Record<string, unknown>> {
        return Promise.resolve({
            outputProperty: {
                $schema: 'http://json-schema.org/draft-04/schema#',
                type: 'object',
                properties: {
                    inputProperty: {
                        type: 'boolean',
                    },
                    updateCount: {
                        type: 'number',
                    },
                },
            },
        });
    }

    /**
     * Used to initialize the control instance. Controls can kick off remote server calls and other initialization actions here.
     * Data-set values are not initialized here, use updateView.
     * @param context The entire property bag available to control via Context Object; It contains values as set up by the customizer mapped to property names defined in the manifest, as well as utility functions.
     * @param notifyOutputChanged A callback method to alert the framework that the control has new outputs ready to be retrieved asynchronously.
     * @param state A piece of data that persists in one session for a single user. Can be set at any point in a controls life cycle by calling 'setControlState' in the Mode interface.
     * @param container If a control is marked control-type='standard', it will receive an empty div element within which it can render its content.
     */
    public init(
        context: ComponentFramework.Context<IInputs>,
        notifyOutputChanged: () => void,
        state: ComponentFramework.Dictionary,
        container: HTMLDivElement,
    ): void {
        this.notifyOutputChanged = notifyOutputChanged;
        // Add control initialization code
        this.container = container;
        this.container.innerHTML = `Updated ${this.updateCount} ${this.updateCount === 1 ? 'time' : 'times'}`;
        this.inputProperty = context.parameters.inputProperty.raw;
        this.boundProperty = context.parameters.boundProperty.raw;
    }

    /**
     * Called when any value in the property bag has changed. This includes field values, data-sets, global values such as container height and width, offline status, control metadata values such as label, visible, etc.
     * @param context The entire property bag available to control via Context Object; It contains values as set up by the customizer mapped to names defined in the manifest, as well as utility functions
     */
    public updateView(context: ComponentFramework.Context<IInputs>): void {
        // Add control initialization code
        this.updateCount += 1;
        this.container.innerHTML = `Updated ${this.updateCount} ${this.updateCount === 1 ? 'time' : 'times'}`;
        this.boundProperty = context.parameters.boundProperty.raw;
        if (this.inputProperty !== context.parameters.inputProperty.raw) {
            this.inputProperty = context.parameters.inputProperty.raw;
            this.notifyOutputChanged();
        }
    }
    
    /**
     * It is called by the framework prior to a control receiving new data.
     * @returns an object based on nomenclature defined in manifest, expecting object[s] for property marked as “bound” or “output”
     */
    public getOutputs(): IOutputs {
        const updateCount = this.updateCount;
        const boundProperty = this.boundProperty;
        const inputProperty = this.inputProperty;
        const outputProperty = { updateCount, inputProperty };
        return { boundProperty, outputProperty };
    }

    /**
     * Called when the control is to be removed from the DOM tree. Controls should use this call for cleanup.
     * i.e. cancelling any pending remote calls, removing listeners, etc.
     */
    public destroy(): void {
        // Add code to cleanup control if necessary
    }
}
