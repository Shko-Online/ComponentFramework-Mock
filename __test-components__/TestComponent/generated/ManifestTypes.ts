/*
*This is auto generated from the ControlManifest.Input.xml file
*/

// Define IInputs and IOutputs Type. They should match with ControlManifest.
export interface IInputs {
    selection: ComponentFramework.PropertyTypes.MultiSelectOptionSetProperty;
    controlType: ComponentFramework.PropertyTypes.EnumProperty<"0" | "1">;
    orientation: ComponentFramework.PropertyTypes.EnumProperty<"1" | "0">;
}
export interface IOutputs {
    selection?: number[];
}
