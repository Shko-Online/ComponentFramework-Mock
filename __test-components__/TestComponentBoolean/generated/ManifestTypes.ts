/*
*This is auto generated from the ControlManifest.Input.xml file
*/

// Define IInputs and IOutputs Type. They should match with ControlManifest.
export interface IInputs {
    turnedOn: ComponentFramework.PropertyTypes.TwoOptionsProperty;
}
export interface IOutputs {
    turnedOn?: boolean;
}
