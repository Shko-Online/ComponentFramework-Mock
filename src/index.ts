/*
Copyright (c) 2022 Betim Beja and Shko Online LLC
Licensed under the MIT license.
*/

export type { MockToRaw, PropertyMap, PropertyToMock } from './ComponentFramework-Mock';
export type { MockGenerator } from './ComponentFramework-Mock-Generator';
export type { ShkoOnline } from './ShkoOnline';

export {
    AttributeMetadataCollection,
    AttributeType,
    ClientMock,
    ContextMock,
    DataSetMock,
    DateTimeMetadataMock,
    DateTimePropertyMock,
    DecimalNumberMetadataMock,
    DecimalNumberPropertyMock,
    DeviceMock,
    EntityRecordMock,
    EnumPropertyMock,
    FactoryMock,
    FilteringMock,
    FormattingMock,
    LinkingMock,
    LookupMetadataMock,
    LookupPropertyMock,
    MetadataMock,
    ModeMock,
    MultiSelectOptionSetPropertyMock,
    NavigationMock,
    NumberMetadataMock,
    NumberPropertyMock,
    OptionMetadataMock,
    OptionSetMetadataMock,
    OptionSetPropertyMock,
    OptionSetType,
    PagingMock,
    PopupServiceMock,
    PropertyMock,
    ResourcesMock,
    StringMetadataMock,
    StringPropertyMock,
    TwoOptionMetadataMock,
    TwoOptionsPropertyMock,
    UserSettingsMock,
    UtilityMock,
    WebApiMock,
    WholeNumberMetadataMock,
    WholeNumberPropertyMock,
} from './ComponentFramework-Mock';

export {
    ComponentFrameworkMockGeneratorReact,
    ComponentFrameworkMockGenerator,
    MetadataDB,
    ReactResizeObserver,
    mockGetEntityMetadata,
    mockRefreshDatasets,
    mockRefreshParameters,
    mockSetControlResource,
    mockSetControlState,
} from './ComponentFramework-Mock-Generator';

export { arrayEqual, AttributeMetadataGenerator, itemEqual, newGuid, showBanner } from './utils';
