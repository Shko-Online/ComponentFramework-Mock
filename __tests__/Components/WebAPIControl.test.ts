/*
    Copyright (c) 2022 Betim Beja and Shko Online LLC
    Licensed under the MIT license.
*/

import { it, expect, describe, beforeEach } from '@jest/globals';
import { WebAPIControl as Control } from '../../__sample-components__/WebAPIControl';
import { IInputs, IOutputs } from '../../__sample-components__/WebAPIControl/generated/ManifestTypes';
import { AttributeType, ComponentFrameworkMockGenerator, ShkoOnline, StringPropertyMock } from '../../src';

jest.useFakeTimers();

describe('WebAPIControl', () => {
    let mockGenerator: ComponentFrameworkMockGenerator<IInputs, IOutputs>;
    beforeEach(() => {
        const container = document.createElement('div');
        mockGenerator = new ComponentFrameworkMockGenerator(
            Control,
            {
                stringProperty: StringPropertyMock
            },
            container,
        );
        mockGenerator.metadata.initMetadata([
            { LogicalName: 'account', EntitySetName: 'accounts', PrimaryIdAttribute: 'accountid', PrimaryNameAttribute: 'name', Attributes: [
                {
                    AttributeType: AttributeType.Uniqueidentifier,
                    LogicalName: 'accountid',
                    SchemaName: 'AccountId'
                } as ShkoOnline.AttributeMetadata,
                {
                    AttributeType: AttributeType.String,
                    LogicalName: 'name',
                    SchemaName: 'Name'
                } as ShkoOnline.StringAttributeMetadata,
                {
                    AttributeType: AttributeType.Money,
                    LogicalName: 'revenue',
                    SchemaName: 'revenue'
                } as ShkoOnline.AttributeMetadata
            ] },
        ]);

        mockGenerator.context._SetCanvasItems({
            stringProperty: '',
        });
        document.body.appendChild(container);
    });

    afterEach(() => {
        document.body.innerHTML = null;
    });

    it('Should render', () => {
        mockGenerator.ExecuteInit();
        mockGenerator.ExecuteUpdateView();
        expect(mockGenerator.container).toMatchSnapshot();
    });

    it('Should create record', async () => {
        mockGenerator.ExecuteInit();
        mockGenerator.ExecuteUpdateView();

        const button = mockGenerator.container.getElementsByTagName('button')[0];
        var evt = document.createEvent('Event');
        evt.initEvent('click', false, true);
        button.dispatchEvent(evt);

        await jest.runAllTimersAsync();

        var result_container = mockGenerator.container.getElementsByClassName('result_container')[0];

        expect(result_container.innerHTML).toContain('Created new account record with below values:');
        // expect(document.body).toMatchSnapshot();
    });

  
});
