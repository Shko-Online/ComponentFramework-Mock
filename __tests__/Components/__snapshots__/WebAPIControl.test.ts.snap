// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebAPIControl Should render 1`] = `
<div>
  <div
    class="WebAPIControl_Container"
  >
    <div
      class="SampleControl_WebAPIControl_Header create_container_header"
    />
    <button
      buttonvalue="100"
      class="SampleControl_WebAPIControl_ButtonClass"
      id="create_button_100"
    >
      Create record with annual revenue of 100
    </button>
    <button
      buttonvalue="200"
      class="SampleControl_WebAPIControl_ButtonClass"
      id="create_button_200"
    >
      Create record with annual revenue of 200
    </button>
    <button
      buttonvalue="300"
      class="SampleControl_WebAPIControl_ButtonClass"
      id="create_button_300"
    >
      Create record with annual revenue of 300
    </button>
    <div
      class="SampleControl_WebAPIControl_Header delete_container_header"
    />
    <button
      class="SampleControl_WebAPIControl_ButtonClass"
      id="delete_button"
    >
      Select record to delete
    </button>
    <div
      class="SampleControl_WebAPIControl_Header fetchxml_status_container_header"
    />
    <div
      class="fetchxml_status_container"
    />
    <button
      class="SampleControl_WebAPIControl_ButtonClass"
      id="odata_refresh"
    >
      Calculate average value of annual revenue
    </button>
    <div
      class="SampleControl_WebAPIControl_Header odata_status_container_header"
    />
    <div
      class="odata_status_container"
    />
    <button
      class="SampleControl_WebAPIControl_ButtonClass"
      id="odata_refresh"
    >
      Refresh record count
    </button>
    <div
      class="SampleControl_WebAPIControl_Header result_container_header"
    />
    <div
      class="result_container"
    >
      Web API sample custom control loaded
    </div>
  </div>
</div>
`;
